File: UI/SPA/React/react_template/src/components/CrudSample.js
import * as React from 'react';
import 'isomorphic-fetch';
import {CrudSampleRootUrl} from '../const.js';

import oauth_oidc from '../touryo/oauth_oidc';

export class CrudSample extends React.Component {
    // constructor
    constructor() {
        super();

        this.state = {
            message : "",
            ddl : {
                ddlDap: "SQL",
                ddlMode1: "individual",
                ddlMode2: "static",
                ddlIso: "NT",
                ddlExRollback: "-",
                ddlOrder: "c1",
                ddlOrderSequence: "A"
            },
            shipper : {
                shipperID: "",
                companyName: "",
                phone: ""
            },
            shippers : 
            [
                {
                    shipperID: "",
                    companyName: "",
                    phone: ""
                }
            ],
            loading: true
        };

        // データアクセス制御クラス
        this.ddlDap = [
            { label: "SQL Server / SQL Client", value: "SQL" }, 
            { label: "Multi-DB / OLEDB.NET", value: "OLE" },
            { label: "Multi-DB / ODBC.NET", value: "ODB" },
            { label: "Oracle / ODP.NET", value: "ODP" },
            { label: "DB2 / DB2.NET", value: "DB2" },
            { label: "HiRDB / HiRDB-DP", value: "HIR" },
            { label: "MySQL Cnn/NET", value: "MCN" },
            { label: "PostgreSQL / Npgsql", value: "NPS" }
        ];
        // Ｄａｏモード
        this.ddlMode1 = [
            { label: "個別Ｄａｏ", value: "individual" }, 
            { label: "共通Ｄａｏ", value: "common" },
            { label: "自動生成Ｄａｏ（更新のみ）", value: "generate" }
        ];
        // 静的、動的のクエリ モード
        this.ddlMode2 = [
            { label: "静的クエリ", value: "static" },
            { label: "動的クエリ", value: "dynamic" }
        ];
        // 分離レベル
        this.ddlIso = [
            { label: "ノットコネクト", value: "NC" },
            { label: "ノートランザクション", value: "NT" },
            { label: "ダーティリード", value: "RU" },
            { label: "リードコミット", value: "RC" },
            { label: "リピータブルリード", value: "RR" },
            { label: "シリアライザブル", value: "SZ" },
            { label: "スナップショット", value: "SS" },
            { label: "デフォルト", value: "DF" }
        ];
        // コミット、ロールバックを設定
        this.ddlExRollback = [
            { label: "正常時", value: "-" },
            { label: "業務例外", value: "Business" },
            { label: "システム例外", value: "System" },
            { label: "その他、一般的な例外", value: "Other" },
            { label: "業務例外への振替", value: "Other-Business" },
            { label: "システム例外への振替", value: "Other-System" } 
        ];
        // 並び替え対象列
        this.ddlOrder = [
            { label: "c1", value: "c1" },
            { label: "c2", value: "c2" },
            { label: "c3", value: "c3" }
        ];
        // 昇順・降順
        this.ddlOrderSequence = [
            { label: "ASC", value: "" },
            { label: "DESC", value: "D" }
        ];

        // method を bind
        // drop down list
        this.onChangeDdlDap = this.onChangeDdlDap.bind(this);
        this.onChangeDdlMode1 = this.onChangeDdlMode1.bind(this);
        this.onChangeDdlMode2 = this.onChangeDdlMode2.bind(this);
        this.onChangeDdlIso = this.onChangeDdlIso.bind(this);
        this.onChangeDdlExRollback = this.onChangeDdlExRollback.bind(this);
        this.onChangeDdlOrder = this.onChangeDdlOrder.bind(this);
        this.onChangeDdlOrderSequence = this.onChangeDdlOrderSequence.bind(this);
        
        // input type text
        this.onChangeShipperID = this.onChangeShipperID.bind(this);
        this.onChangeCompanyName = this.onChangeCompanyName.bind(this);
        this.onChangePhone = this.onChangePhone.bind(this);
    }

    // render
    render() {

        const div0Style = { };

        const div1Style = {
            display : 'inline-block'
        };

        const div2Style = {
            display : 'inline-block'
        };

        let contents = null;
        if(this.state.loading)
        {
            contents = <p><em>...Table...</em></p>;
        }
        else
        {
            contents = this.renderTable(this.state.shippers); 
        }

        return <div>
            <h1>CRUD sample</h1>
            <p>This component demonstrates CRUD.</p>
            <div style={div0Style}>
                { this.renderDDL() }
            </div>
            <div style={div1Style}>
                { this.renderInput() }
            </div>
            <div style={div2Style}>
                { contents }
                <p>処理結果：{this.state.message}</p>
            </div>
            <div>
                <button className='btn' onClick={ () => { this.selectCount() } }>SelectCount</button>&nbsp;
                <button className='btn' onClick={ () => { this.selectAll_DT() } }>SelectAll_DT</button>&nbsp;
                <button className='btn' onClick={ () => { this.selectAll_DS() } }>SelectAll_DS</button>&nbsp;
                <button className='btn' onClick={ () => { this.selectAll_DR() } }>SelectAll_DR</button>&nbsp;
                <button className='btn' onClick={ () => { this.selectAll_DSQL() } }>SelectAll_DSQL</button>&nbsp;
                <button className='btn' onClick={ () => { this.select() } }>Select</button>&nbsp;
                <button className='btn' onClick={ () => { this.insert() } }>Insert</button>&nbsp;
                <button className='btn' onClick={ () => { this.update() } }>Update</button>&nbsp;
                <button className='btn' onClick={ () => { this.delete() } }>Delete</button>
            </div>
        </div>;
    }
    renderDDL()
    {
        const ddlStyle = {
            width:"100%"
        };

        return <table className='table'>
            <tbody style={ddlStyle}>
                <tr>
                    <td>データアクセス制御クラス:</td>
                    <td>
                        <select value={this.state.ddl.ddlDap}
                            defaultValue={this.state.ddl.ddlDap} onChange={this.onChangeDdlDap}>
                            { this.ddlDap.map( d => <option key={d.value} value={d.value}>{d.label}</option>)}
                        </select>
                    </td>
                </tr>
                <tr>
                    <td>Ｄａｏモード:</td>
                    <td>
                        <select value={this.state.ddl.ddlMode1}
                            defaultValue={this.state.ddl.ddlMode1} onChange={this.onChangeDdlMode1}>
                            { this.ddlMode1.map( d => <option key={d.value} value={d.value}>{d.label}</option>)}
                        </select><br/>
                    </td>
                </tr>
                <tr>
                    <td>静的、動的のクエリ モード:</td>
                    <td>
                        <select value={this.state.ddl.ddlMode2}
                            defaultValue={this.state.ddl.ddlMode2} onChange={this.onChangeDdlMode2}>
                            { this.ddlMode2.map( d => <option key={d.value} value={d.value}>{d.label}</option>)}
                        </select>
                    </td>
                </tr>
                <tr>
                    <td>分離レベル:</td>
                    <td>
                        <select value={this.state.ddl.ddlIso}
                            defaultValue={this.state.ddl.ddlIso} onChange={this.onChangeDdlIso}>
                            { this.ddlIso.map( d => <option key={d.value} value={d.value}>{d.label}</option>)}
                        </select>
                    </td>
                </tr>
                <tr>
                    <td>コミット、ロールバックを設定:</td>
                    <td>
                        <select value={this.state.ddl.ddlExRollback}
                            defaultValue={this.state.ddl.ddlExRollback} onChange={this.onChangeDdlExRollback}>
                            { this.ddlExRollback.map( d => <option key={d.value} value={d.value}>{d.label}</option>)}
                        </select>
                    </td>
                </tr>
                <tr>
                    <td>並び替え対象列:</td>
                    <td>
                        <select value={this.state.ddl.ddlOrder}
                            defaultValue={this.state.ddl.ddlOrder} onChange={this.onChangeDdlOrder}>
                            { this.ddlOrder.map( d => <option key={d.value} value={d.value}>{d.label}</option>)}
                        </select>
                    </td>
                </tr>
                <tr>
                    <td>昇順・降順:</td>
                    <td>
                        <select value={this.state.ddl.ddlOrderSequence}
                            defaultValue={this.state.ddl.ddlOrderSequence} onChange={this.onChangeDdlOrderSequence}>
                            { this.ddlOrderSequence.map( d => <option key={d.value} value={d.value}>{d.label}</option>)}
                        </select>
                    </td>
                </tr>
            </tbody>
        </table>;
    }
    renderInput()
    {
        const inputStyle = {
            width:"100%"
        };

        return <table className='table'>
            <tbody style={inputStyle}>
                <tr>
                    <td>ShipperID：</td>
                    <td>
                        <p>
                            <label>
                                <input type="text" id="txtShipperID" value={this.state.shipper.shipperID} onChange={this.onChangeShipperID}/>
                            </label>
                        </p>            
                    </td>
                </tr>
                <tr>
                    <td>CompanyName：</td>
                    <td>
                        <p>
                            <label>
                                <input type="text" id="txtCompanyName" value={this.state.shipper.companyName} onChange={this.onChangeCompanyName}/>
                            </label>
                        </p>
                    </td>
                </tr>
                <tr>
                    <td>Phone：</td>
                    <td>
                        <p>
                            <label>
                                <input type="text" id="txtPhone" value={this.state.shipper.phone} onChange={this.onChangePhone}/>
                            </label>
                        </p>
                    </td>
                </tr>
            </tbody>
        </table>;
    }
    renderTable(shippers) {
        return <table className='table'>
            <thead>
                <tr>
                    <th>ShipperID</th>
                    <th>CompanyName</th>
                    <th>Phone</th>
                </tr>
            </thead>
            <tbody>
            {shippers.map(shipper =>
                <tr key={ shipper.shipperID }>
                    <td>{ shipper.shipperID }</td>
                    <td>{ shipper.companyName }</td>
                    <td>{ shipper.phone }</td>
                </tr>
            )}
            </tbody>
        </table>;
    }

    // event handler (JavaScript)

    // drop down list
    onChangeDdlDap(e){
        this.setState({ddl: { ...this.state.ddl, ddlDap: e.target.value }});
    }
    onChangeDdlMode1(e){
        this.setState({ddl: { ...this.state.ddl, ddlMode1: e.target.value }});
    }
    onChangeDdlMode2(e){
        this.setState({ddl: { ...this.state.ddl, ddlMode2: e.target.value }});
    }
    onChangeDdlIso(e){
        this.setState({ddl: { ...this.state.ddl, ddlIso: e.target.value }});
    }
    onChangeDdlExRollback(e){
        this.setState({ddl: { ...this.state.ddl, ddlExRollback: e.target.value }});
    }
    onChangeDdlOrder(e){
        this.setState({ddl: { ...this.state.ddl, ddlOrder: e.target.value }});
    }
    onChangeDdlOrderSequence(e){
        this.setState({ddl: { ...this.state.ddl, ddlOrderSequence: e.target.value }});
    }

    // input type text
    onChangeShipperID(e){
        this.setState({shipper: { ...this.state.shipper, shipperID: e.target.value }});
    }
    onChangeCompanyName(e){
        this.setState({shipper: { ...this.state.shipper, companyName: e.target.value }});
    }
    onChangePhone(e){
        this.setState({shipper: { ...this.state.shipper, phone: e.target.value }});
    }

    // event handler (WebAPI)
    selectCount() {
        // エラーメッセージをクリアする
        this.setState({
            message : ""
        });

        // リクエストの生成
        const method = "POST";
        const headers = createHttpRequestHeader(false);
        const body = 
            "ddlDap=" + this.state.ddl.ddlDap
            + "&ddlMode1=" + this.state.ddl.ddlMode1
            + "&ddlMode2=" + this.state.ddl.ddlMode2
            + "&ddlExRollback=" + this.state.ddl.ddlExRollback;

        // fetchする。
        fetch(CrudSampleRootUrl + 'SelectCount', {method, headers, body})
        .then(fetchStatusHandler) 
        .then(response => response.json())
        .then(data => {
            if(data.message) {
                this.setState({ message : data.message });
            }
            else if(data.errorMSG) {
                this.setState({ message : JSON.stringify(data.errorMSG) });
            }
            else if(data.exceptionMSG) {
                this.setState({ message : JSON.stringify(data.exceptionMSG) });
            }
        })
        .catch(
            // 異常系
            error => {
                this.setState({ message : JSON.stringify(error.stack) });
            }
        );
    }
    selectAll_DT() {
        // エラーメッセージをクリアする
        this.setState({
            message : ""
        });

        // リクエストの生成
        const method = "POST";
        const headers = createHttpRequestHeader(false);
        const body = 
            "ddlDap=" + this.state.ddl.ddlDap
            + "&ddlMode1=" + this.state.ddl.ddlMode1
            + "&ddlMode2=" + this.state.ddl.ddlMode2
            + "&ddlExRollback=" + this.state.ddl.ddlExRollback;

        // fetchする。
        fetch(CrudSampleRootUrl + 'SelectAll_DT', {method, headers, body})
        .then(fetchStatusHandler)     
        .then(response => response.json())
        .then(data => {
            if(data.result) {
                this.setState({
                    message : "",
                    shippers : data.result,
                    loading: false
                });
            }
            else if(data.errorMSG) {
                this.setState({ message : JSON.stringify(data.errorMSG) });
            }
            else if(data.exceptionMSG) {
                this.setState({ message : JSON.stringify(data.exceptionMSG) });
            }
        })
        .catch(
            // 異常系
            error => {
                this.setState({ message : JSON.stringify(error.stack) });
            }
        );
    }
    selectAll_DS() {
        // エラーメッセージをクリアする
        this.setState({
            message : ""
        });

        // リクエストの生成
        const method = "POST";
        const headers = createHttpRequestHeader(false);
        const body = 
            "ddlDap=" + this.state.ddl.ddlDap
            + "&ddlMode1=" + this.state.ddl.ddlMode1
            + "&ddlMode2=" + this.state.ddl.ddlMode2
            + "&ddlExRollback=" + this.state.ddl.ddlExRollback;

        // fetchする。
        fetch(CrudSampleRootUrl + 'SelectAll_DS', {method, headers, body})
        .then(fetchStatusHandler) 
        .then(response => response.json())
        .then(data => {
            if(data.result) {
                this.setState({
                    message : "",
                    shippers : data.result,
                    loading: false
                });
            }
            else if(data.errorMSG) {
                this.setState({ message : JSON.stringify(data.errorMSG) });
            }
            else if(data.exceptionMSG) {
                this.setState({ message : JSON.stringify(data.exceptionMSG) });
            }
        })
        .catch(
            // 異常系
            error => {
                this.setState({ message : JSON.stringify(error.stack) });
            }
        );
    }
    selectAll_DR() {
        // エラーメッセージをクリアする
        this.setState({
            message : ""
        });

        // リクエストの生成
        const method = "POST";
        const headers = createHttpRequestHeader(false);
        const body = 
            "ddlDap=" + this.state.ddl.ddlDap
            + "&ddlMode1=" + this.state.ddl.ddlMode1
            + "&ddlMode2=" + this.state.ddl.ddlMode2
            + "&ddlExRollback=" + this.state.ddl.ddlExRollback;

        // fetchする。
        fetch(CrudSampleRootUrl + 'SelectAll_DR', {method, headers, body})
        .then(fetchStatusHandler) 
        .then(response => response.json())
        .then(data => {
            if(data.result) {
                this.setState({
                    message : "",
                    shippers : data.result,
                    loading: false
                });
            }
            else if(data.errorMSG) {
                this.setState({ message : JSON.stringify(data.errorMSG) });
            }
            else if(data.exceptionMSG) {
                this.setState({ message : JSON.stringify(data.exceptionMSG) });
            }
        })
        .catch(
            // 異常系
            error => {
                this.setState({ message : JSON.stringify(error.stack) });
            }
        );
    }
    selectAll_DSQL() {
        // エラーメッセージをクリアする
        this.setState({
            message : ""
        });

        // リクエストの生成
        const method = "POST";
        const headers = createHttpRequestHeader(false);
        const body = 
            "ddlDap=" + this.state.ddl.ddlDap
            + "&ddlMode1=" + this.state.ddl.ddlMode1
            + "&ddlMode2=" + this.state.ddl.ddlMode2
            + "&ddlExRollback=" + this.state.ddl.ddlExRollback
            + "&orderColumn=" + this.state.ddl.ddlOrder
            + "&orderSequence=" + this.state.ddl.ddlOrderSequence;

        // fetchする。
        fetch(CrudSampleRootUrl + 'SelectAll_DSQL', {method, headers, body})
        .then(fetchStatusHandler) 
        .then(response => response.json())
        .then(data => {
            if(data.result) {
                this.setState({
                    message : "",
                    shippers : data.result,
                    loading: false
                });
            }
            else if(data.errorMSG) {
                this.setState({ message : JSON.stringify(data.errorMSG) });
            }
            else if(data.exceptionMSG) {
                this.setState({ message : JSON.stringify(data.exceptionMSG) });
            }
        })
        .catch(
            // 異常系
            error => {
                this.setState({ message : JSON.stringify(error.stack) });
            }
        );
    }
    select() {
        // エラーメッセージをクリアする
        this.setState({
            message : ""
        });

        // リクエストの生成
        const method = "POST";
        const headers = createHttpRequestHeader(true);
        const body = JSON.stringify({
            ddlDap: this.state.ddl.ddlDap,
            ddlMode1: this.state.ddl.ddlMode1,
            ddlMode2: this.state.ddl.ddlMode2,
            ddlExRollback: this.state.ddl.ddlExRollback,
            shipper: {
                shipperID: this.state.shipper.shipperID,
                companyName: "",
                phone: ""
            }
        });

        // fetchする。
        fetch(CrudSampleRootUrl + 'Select', {method, headers, body})
        .then(fetchStatusHandler) 
        .then(response => response.json())
        .then(data => {
            if(data.result) {
                this.setState({
                    shipper:{
                        shipperID: data.result.shipperID,
                        companyName: data.result.companyName,
                        phone: data.result.phone
                        }
                });
            }
            else if(data.errorMSG) {
                this.setState({ message : JSON.stringify(data.errorMSG) });
            }
            else if(data.exceptionMSG) {
                this.setState({ message : JSON.stringify(data.exceptionMSG) });
            }
        })
        .catch(
            // 異常系
            error => {
                this.setState({ message : JSON.stringify(error.stack) });
            }
        );
    }
    insert() {
        // エラーメッセージをクリアする
        this.setState({
            message : ""
        });

        // リクエストの生成
        const method = "POST";
        const headers = createHttpRequestHeader(true);
        const body = JSON.stringify({
            ddlDap: this.state.ddl.ddlDap,
            ddlMode1: this.state.ddl.ddlMode1,
            ddlMode2: this.state.ddl.ddlMode2,
            ddlExRollback: this.state.ddl.ddlExRollback,
            shipper: {
                shipperID: '0',
                companyName: this.state.shipper.companyName,
                phone: this.state.shipper.phone
            }
        });

        // fetchする。
        fetch(CrudSampleRootUrl + 'Insert', {method, headers, body})
        .then(fetchStatusHandler)   
        .then(response => response.json())
        .then(data =>  {
            if(data.message) {
                this.setState({ message : data.message });
            }
            else if(data.errorMSG) {
                this.setState({ message : JSON.stringify(data.errorMSG) });
            }
            else if(data.exceptionMSG) {
                this.setState({ message : JSON.stringify(data.exceptionMSG) });
            }
        })
        .catch(
            // 異常系
            error => {
                this.setState({ message : JSON.stringify(error.stack) });
            }
        );
    }
    update() {
        // エラーメッセージをクリアする
        this.setState({
            message : ""
        });

        // リクエストの生成
        const method = "POST";
        const headers = createHttpRequestHeader(true);
        const body = JSON.stringify({
            ddlDap: this.state.ddl.ddlDap,
            ddlMode1: this.state.ddl.ddlMode1,
            ddlMode2: this.state.ddl.ddlMode2,
            ddlExRollback: this.state.ddl.ddlExRollback,
            shipper: {
                shipperID: this.state.shipper.shipperID,
                companyName: this.state.shipper.companyName,
                phone: this.state.shipper.phone
            }
        });

        // fetchする。
        fetch(CrudSampleRootUrl + 'Update', {method, headers, body})
        .then(fetchStatusHandler)
        .then(response => response.json())
        .then(data => {
            if(data.message) {
                this.setState({ message : data.message });
            }
            else if(data.errorMSG) {
                this.setState({ message : JSON.stringify(data.errorMSG) });
            }
            else if(data.exceptionMSG) {
                this.setState({ message : JSON.stringify(data.exceptionMSG) });
            }
        })
        .catch(
            // 異常系
            error => {
                this.setState({ message : JSON.stringify(error.stack) });
            }
        );
    }
    delete() {
        // エラーメッセージをクリアする
        this.setState({
            message : ""
        });

        // リクエストの生成
        const method = "POST";
        const headers = createHttpRequestHeader(true);
        const body = JSON.stringify({
            ddlDap: this.state.ddl.ddlDap,
            ddlMode1: this.state.ddl.ddlMode1,
            ddlMode2: this.state.ddl.ddlMode2,
            ddlExRollback: this.state.ddl.ddlExRollback,
            shipper: {
                shipperID: this.state.shipper.shipperID,
                companyName: "",
                phone: ""
            }
        });

        // fetchする。
        fetch(CrudSampleRootUrl + 'Delete', {method, headers, body})
        .then(fetchStatusHandler)
        .then(response => response.json())
        .then(data => {
            if(data.message) {
                this.setState({ message : data.message });
            }
            else if(data.errorMSG) {
                this.setState({ message : JSON.stringify(data.errorMSG) });
            }
            else if(data.exceptionMSG) {
                this.setState({ message : JSON.stringify(data.exceptionMSG) });
            }
        })
        .catch(
            // 異常系
            error => {
                this.setState({ message : JSON.stringify(error.stack) });
            }
        );
    }
}

function createHttpRequestHeader(isJsonRpc) {
  let headers = { };

  if(isJsonRpc) {
    headers = {
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    };
  }
  else {
    headers = {
      'Accept': 'application/json',
      'Content-Type': 'application/x-www-form-urlencoded'
    };
  }

  let access_token = oauth_oidc.getAccessToken();
  if(access_token) {
    headers.Authorization = "Bearer " + access_token;
  }
  
  return headers;
}

// https://github.com/github/fetch/issues/155#issuecomment-108353192
function fetchStatusHandler(response) {
  if (response.status === 200) {
      return response;
  }
  else {
      throw new Error(response.statusText);
  }
}

File: UI/XPlat/flutter_template/lib/screens/crud_sample.dart
import '../importer.dart';

class ScreenCrudSample extends StatefulWidget {
  const ScreenCrudSample({super.key, required this.title});

  final String title;

  @override
  State<ScreenCrudSample> createState() => _CrudSampleState();
}

class _CrudSampleState extends State<ScreenCrudSample> {
  // ddl値
  String _ddlDap = "SQL";
  String _ddlMode1 = "individual";
  String _ddlMode2 = "static";
  String _ddlIso = "RC";
  String _ddlExRollback = "-";
  String _orderColumn = "c1";
  String _orderSequence = "";

  // 値取得
  final _formKey = GlobalKey<FormState>();
  // 値設定
  final _shipperIDKey = GlobalKey<FormFieldState>();
  final _companyNameKey = GlobalKey<FormFieldState>();
  final _phoneKey = GlobalKey<FormFieldState>();
  // Field値
  String _shipperID = "";
  String _companyName = "";
  String _phone = "";
  // フォーカス制御
  //final _shipperIDFocusNode = FocusNode();
  final _companyNameFocusNode = FocusNode();
  final _phoneFocusNode = FocusNode();

  // JSON値
  String _display = "";
  List<dynamic> _jsonItems  = jsonDecode(
      '[{"shipperID":"shipperID","companyName":"companyName","phone":"phone"}]'
  );

  @override
  void initState() {
    super.initState();
  }

  Future<void> _selectCount() async {
    final url = Uri.parse('${AppConfig.crudSampleRootUrl}SelectCount').toString();
    final headers = await createHttpRequestHeader();

    final body = {
      "ddlDap": _ddlDap,
      "ddlMode1": _ddlMode1,
      "ddlMode2": _ddlMode2,
      "ddlExRollback": _ddlExRollback,
    };

    await postFetch(
      url,
      headers,
      body,
      (data) {
        setState(() {
          _display = (data as Map<String, dynamic>)['message'];
        });
      },
      (msg) {
        setState(() {
          _display = msg;
        });
      },
    );
  }

  Future<void> _selectAll(String urlPrefix) async {
    final url = Uri.parse('${AppConfig.crudSampleRootUrl}SelectAll_$urlPrefix').toString();
    final headers = await createHttpRequestHeader();

    await postFetch(
      url,
      headers,
      {
        "ddlDap": _ddlDap,
        "ddlMode1": _ddlMode1,
        "ddlMode2": _ddlMode2,
        "ddlExRollback": _ddlExRollback,
        "orderColumn": _orderColumn,
        "orderSequence": _orderSequence,
      },
      (data) {
        setState(() {
          _display = data['message'];
          _jsonItems = data['result'];
        });
      },
      (msg) {
        setState(() {
          _display = msg;
        });
      },
    );
  }

  Future<void> _crud(String urlPrefix) async {
    _formKey.currentState?.save();
    final url = Uri.parse('${AppConfig.crudSampleRootUrl}$urlPrefix').toString();

    final headers = await createHttpRequestHeader(isJsonRpc: true);

    final body = jsonEncode({
      "ddlDap": _ddlDap,
      "ddlMode1": _ddlMode1,
      "ddlMode2": _ddlMode2,
      "ddlExRollback": _ddlExRollback,
      "shipper": {
        "shipperID": _shipperID,
        "companyName": _companyName,
        "phone": _phone,
      },
    });

    await postFetch(
      url,
      headers,
      body,
      (data) {
        setState(() {
          _display = data['message'];
          if (urlPrefix == "Select") {
            final shipper = data['result'];
            _shipperIDKey.currentState?.didChange(shipper['shipperID']);
            _companyNameKey.currentState?.didChange(shipper['companyName']);
            _phoneKey.currentState?.didChange(shipper['phone']);
          }
        });
      },
      (msg) {
        setState(() {
          _display = msg;
        });
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SingleChildScrollView(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: <Widget>[
            Row(
              children: <Widget>[
                Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: <Widget>[
                    MyDropdownButton(
                      "ddlDap",
                      (value) {
                        setState(() {
                          _ddlDap = value.toString();
                        });
                      },
                      _ddlDap,
                      {
                        "SQL Server / SQL Client" : "SQL",
                        "Multi-DB / OLEDB.NET" : "OLE",
                        "Multi-DB / ODBC.NET" : "ODB",
                        "Oracle / ODP.NET" : "ODP",
                        "DB2 / DB2.NET" : "DB2",
                        "HiRDB / HiRDB-DP" : "HIR",
                        "MySQL Cnn/NET" : "MCN",
                        "PostgreSQL / Npgsql" : "NPS"
                      }
                    ),
                    MyDropdownButton(
                      "ddlMode1",
                      (value) {
                        setState(() {
                          _ddlMode1 = value.toString();
                        });
                      },
                      _ddlMode1,
                      {
                        "個別Ｄａｏ" : "individual",
                        "共通Ｄａｏ" : "common",
                        "自動生成Ｄａｏ（更新のみ）" : "generate",
                      }
                    ),
                    MyDropdownButton(
                      "ddlMode2",
                        (value) {
                          setState(() {
                            _ddlMode2 = value.toString();
                        });
                      },
                      _ddlMode2,
                      {
                        "静的クエリ" : "static",
                        "動的クエリ" : "dynamic",
                      }
                    ),
                    MyDropdownButton(
                      "ddlIso",
                      (value) {
                        setState(() {
                          _ddlIso = value.toString();
                        });
                      },
                      _ddlIso,
                      {
                        "ノットコネクト" : "NC",
                        "ノートランザクション" : "NT",
                        "ダーティリード" : "RU",
                        "リードコミット" : "RC",
                        "リピータブルリード" : "RR",
                        "シリアライザブル" : "SZ",
                        "スナップショット" : "SS",
                        "デフォルト" : "DF",
                      }
                    ),
                    MyDropdownButton(
                      "ddlExRollback",
                      (value) {
                        setState(() {
                          _ddlExRollback = value.toString();
                        });
                      },
                      _ddlExRollback,
                      {
                        "正常時" : "-",
                        "業務例外" : "Business",
                        "システム例外" : "System",
                        "その他、一般的な例外" : "Other",
                        "業務例外への振替" : "Other-Business",
                        "システム例外への振替" : "Other-System",
                      }
                    ),
                    MyDropdownButton(
                      "ddlOrder",
                      (value) {
                        setState(() {
                          _orderColumn = value.toString();
                        });
                      },
                      _orderColumn,
                      {
                        "c1" : "c1",
                        "c2" : "c2",
                        "c3" : "c3",
                      }
                    ),
                    MyDropdownButton(
                      "ddlOrderSequence",
                      (value) {
                        setState(() {
                          _orderSequence = value.toString();
                        });
                      },
                      _orderSequence,
                      {
                        "ASC" : "",
                        "DESC" : "D",
                      }
                    )
                  ],
                ),
                Expanded(
                  child: Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    crossAxisAlignment: CrossAxisAlignment.center,
                    children: <Widget>[
                      DataTable(
                        columns: [
                          DataColumn(label: Text('shipperID')),
                          DataColumn(label: Text('companyName')),
                          DataColumn(label: Text('phone')),
                        ],
                        rows: (_jsonItems).map((element) => DataRow(
                          cells: <DataCell>[
                            DataCell(Text(element["shipperID"] ?? "")),
                            DataCell(Text(element["companyName"] ?? "")),
                            DataCell(Text(element["phone"] ?? "")),
                          ]
                        )).toList(),
                      ),
                    ]
                  ),
                ),
              ],                            
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Text('件数:'),
                Text(_display),
              ],
            ),
            Form(
              key: _formKey,
              child: Column(
                children: [
                  TextFormField(
                    key: _shipperIDKey,
                    decoration: InputDecoration(labelText: 'shipperID'),
                    textInputAction: TextInputAction.next,
                    onFieldSubmitted: (_) {
                      FocusScope.of(context).requestFocus(_companyNameFocusNode);
                    },
                    onSaved: (value) {
                      _shipperID = value ?? "";
                    },
                  ),
                  TextFormField(
                    key: _companyNameKey,
                    decoration: InputDecoration(labelText: 'companyName'),
                    textInputAction: TextInputAction.next,
                    focusNode: _companyNameFocusNode,
                    onFieldSubmitted: (_) {
                      FocusScope.of(context).requestFocus(_companyNameFocusNode);
                    },
                    onSaved: (value) {
                      _companyName = value ?? "";
                    },
                  ),
                  TextFormField(
                    key: _phoneKey,
                    decoration: InputDecoration(labelText: 'phone'),
                    focusNode: _phoneFocusNode,
                    onSaved: (value) {
                      _phone = value ?? "";
                    },
                  ),
                ],
              ),
            ),
            Row(
                mainAxisAlignment: MainAxisAlignment.spaceAround,
                children: <Widget>[
                  MyElevatedButton('Select', (){ _crud("Select"); }),
                  MyElevatedButton('Insert', (){ _crud("Insert"); }),
                  MyElevatedButton('Update', (){ _crud("Update"); }),
                  MyElevatedButton('Delete', (){ _crud("Delete"); }),
                  MyElevatedButton('SelectCount', _selectCount),
                  MyElevatedButton('SelectAll_DT', (){ _selectAll("DT"); }),
                  MyElevatedButton('SelectAll_DSQL', (){ _selectAll("DSQL"); }),
                ]
            ),          
          ],
        )
      ),
    );
  }
}